doc: add documentation for PDN management (#124)
NordicPlayground/nrfxlib: bsdlib/doc/pdn_extension.rst

   To fix this error, you must create another PDN socket that is associated with the APN, and bind the data socket to it.

Configuring the type of transport
*********************************

.. code-block:: c

   nrf_setsockopt(pdn_fd, NRF_SOL_PDN, NRF_SO_PDN_AF, [NRF_AF_INET, NRF_AF_INET6],
                  sizeof(nrf_sa_family_t) * 2);

You can configure a PDN socket to use a specific internet family type of transport, for example, IPv4 only, IPv6 only, or dual stack.
To do so, set the socket option NRF_SO_PDN_AF to request a specific type of connection.
You must do this after creating the socket and before activating it with ``nrf_connect``.

To retrieve a list of what was activated by the network, call ``getsockopt`` again after activation, using the same socket option.

.. admonition:: Limitation

   The current implementation is hardcoded to always request a dual IP stack.

Checking if a PDN is active
***************************

.. code-block:: c

   nrf_getsockopt(pdn_fd, NRF_SOL_PDN, NRF_SO_PDN_STATE, &state, sizeof(nrf_pdn_state_t));

If a data socket is bound to an APN/PDN that is closed, the socket fails.

> Therefore, you should always check if the PDN socket is still active before you use the data socket.

Author: Changed to:
```
If a data socket is bound to an APN/PDN that is closed, the socket fails with an ``NRF_ENETDOWN`` error.

The example code shows how to check whether a PDN socket is still active.
However, if there is a long time...
```

Is that better?

.. _pdn_extension:

PDN management
##############

The BSD library supports configuring up to 10 packet data network (PDN) sockets.
PDN sockets are an extension to the standard socket API.
They make it easy to configure PDN connections as identified by the user's Access Point Name (APN).

After a PDN socket is configured, it can be used by a data socket (IP or RAW socket).
The PDN socket dictates where to send the data, or to do DNS queries on the PDN that is associated with the given APN.

The following sections give pseudo code for typical use cases of PDN sockets.

Creating a PDN socket
*********************

.. code-block:: c

   pdn_fd = nrf_socket(NRF_AF_LTE, NRF_SOCK_MGMT, NRF_NPROTO_PDN);
   nrf_connect(pdn_fd, (struct nrf_sockaddr *)"custom.apn", strlen("custom.apn"));

> When creating a PDN socket, the library requests a context ID (CID) to be created.

Reviewer: Suggestion: move these 3 lines before the code snippet

It then configures the CID, which creates a PDN.
This PDN is used to handle data that is sent over the connection.

> The example code shows how to create a PDN socket with the protocol family NRF_NPROTO_PDN and the proprietary address family NRF_AF_LTE.

Reviewer: This explains the code so it would be better if it immediately followed the snippet (see previous comment)

When the ``nrf_socket`` function returns, a CID has been allocated.

> Next, the created CID must be configured to use the application's APN (using ``nrf_connect``).

Reviewer: ```suggestion
Next, the ``nrf_connect`` function is used to configure the created CID to use the application's APN.
```

If this operation is successful, the library attempts to activate the associated PDN handle.
If this activation succeeds, ``nrf_connect`` returns with success.

The file descriptor that is returned can only be used to operate on the PDN socket itself.
For example, you can use it to set and retrieve socket options, to connect, or to close the PDN socket.

> To use the configured PDN, you must use the associated APN name to identify which PDN to use to transfer data (IP/RAW/DNS data); see `Binding a data socket to an APN`_.

Reviewer: no semicolon but a new sentence 'For details, see ...'

Closing a PDN socket
********************

.. code-block:: c

   nrf_close(pdn_fd);

> The close function for PDN sockets de-registers the associated CID/PDN by deactivating the PDN and removes the configuration of the CID.

Reviewer: ```suggestion
The close function for PDN sockets deregisters the associated CID/PDN by deactivating the PDN.
It also removes the configuration of the CID.
```

Reviewer: It doesn't seem right. I understand deregister doesn't appear in Merriam Webster but it does in some other dictionaries, while de-register is hardly found anywhere. But ok.

The close function for PDN sockets de-registers the associated CID/PDN by deactivating the PDN.
It also removes the configuration of the CID.

Any socket that has been bound to the APN/PDN must be closed before closing the PDN socket.
Otherwise, it will not have any transport and will start reporting errors.

Binding a data socket to an APN
*******************************

.. code-block:: c

   tcp_fd = nrf_socket(NRF_AF_INET, NRF_SOCK_STREAM, NRF_IPPROTO_TCP);
   nrf_setsockopt(tcp_fd, NRF_SOL_SOCKET, NRF_SO_BINDTODEVICE, "custom.apn", strlen("custom.apn"));

>    nrf_bind(...);

Reviewer 2: This line is incorrect, we should remove it.

Author: OK, removed. :)

   nrf_send(tcp_fd, "data", 4, 0);

The example code shows how to create an IPv4 TCP stream socket and bind it to an APN using the NRF_SO_BINDTODEVICE socket option.

> The APN name must be set up and activated prior to binding it; see `Creating a PDN socket`_.

Reviewer: comma or new sentence instead of semicolon

.. note::

>    If the PDN that the socket is being bound to has been deactivated by the network, setting the NRF_SO_BINDTODEVICE option returns an error that the APN parameter is not valid.

Reviewer: ```suggestion
   If the PDN that the socket is being bound to has been deactivated by the network, setting the NRF_SO_BINDTODEVICE option returns an error stating that the APN parameter is not valid.
```

   To fix this error, you must create another PDN socket that is associated with the APN, and bind the data socket to it.

Configuring the type of transport
*********************************

.. code-block:: c

   nrf_setsockopt(pdn_fd, NRF_SOL_PDN, NRF_SO_PDN_AF, [NRF_AF_INET, NRF_AF_INET6],
                  sizeof(nrf_sa_family_t) * 2);

> You can configure a PDN socket to use a specific internet family type of transport, for example, IPv4 only, IPv6 only, or dual stack.

Reviewer: ```suggestion
You can configure a PDN socket to use a specific Internet family type of transport, for example, IPv4 only, IPv6 only, or dual stack.
```

To do so, set the socket option NRF_SO_PDN_AF to request a specific type of connection.
You must do this after creating the socket and before activating it with ``nrf_connect``.

> To retrieve a list of what was activated by the network, call ``getsockopt`` again after activation, using the same socket option.

Reviewer: 'what' sounds cryptic here, does it refer to the activated sockets? not sure
```suggestion
To retrieve a list of sockets that have been activated by the network, call ``getsockopt`` again after activation, using the same socket option.
```

Reviewer 2: You are correct @ru-fu, it refers to the list of address families for the given PDN socket.
That is, a "list" (pointer plus length) of `nrf_socket_family_t` structs.

Author: Updated to "To retrieve a list of what address families were activated by the network..."

.. admonition:: Limitation

   The current implementation is hardcoded to always request a dual IP stack.

Checking if a PDN is active
***************************

.. code-block:: c

   nrf_pdn_state_t state;
   nrf_getsockopt(pdn_fd, NRF_SOL_PDN, NRF_SO_PDN_STATE, &state, sizeof(nrf_pdn_state_t));

If a data socket is bound to an APN/PDN that is closed, the socket fails with an ``NRF_ENETDOWN`` error.

The example code shows how to check whether a PDN socket is still active.
However, if there is a long time between transmissions, the PDN socket might appear to still be active, even though the network has deactivated the PDN.
Only checking the state of the PDN by querying the socket option NRF_SO_PDN_STATE is therefore not reliable.
Instead, send a transmission on the data socket to trigger the link to become active again.

> The network will then report if the PDN has been deactivated.

Reviewer: ```suggestion
The network will then report whether the PDN has been deactivated.
```

Reviewer 2: You are correct @ru-fu.

Author: Leaving as is then.

> In that case, you must create another PDN socket.

Reviewer: ```suggestion
If it has been deactivated, you must create another PDN socket.
```
plus maybe a link back to the 'Creating' section?

.. note::

>    Setting the socket option NRF_SO_PDN_STATE with ``nrf_setsockopt`` to manually control activation/deactivation not supported.

Reviewer: ```suggestion
   Setting the socket option NRF_SO_PDN_STATE with ``nrf_setsockopt`` to manually control activation/deactivation is not supported.
```

Retrieving the context ID (CID)
*******************************

.. code-block:: c

   nrf_pdn_context_id_t cid;
   nrf_getsockopt(pdn_fd, NRF_SOL_PDN, NRF_SO_PDN_CONTEXT_ID, &cid, sizeof(nrf_pdn_context_id_t));

> In some occasions, it is useful to know the CID that is assigned to a PDN socket.

Reviewer: ```suggestion
In some use cases, it is useful to know the CID that is assigned to a PDN socket.
```

*******************

.. code-block:: c

   struct nrf_addrinfo hints = {
     .ai_family = NRF_AF_INET,
     .ai_socktype = NRF_SOCK_STREAM
   }

   struct nrf_addrinfo apn_hints;

   apn_hints.ai_family = NRF_AF_LTE;
   apn_hints.ai_socktype = NRF_SOCK_MGMT;
   apn_hints.ai_protocol = NRF_PROTO_PDN;
   apn_hints.ai_canonname = "custom.apn";

   hints.ai_next = &apn_hints;

   nrf_getaddrinfo("example.com", NULL, &hints, &result);

The example code shows how to specify an APN to use for a DNS query.
In this example, the DNS hints are extended with an extra linked-list hint that specifies the APN to use for the query.

Like for data sockets, the APN name must be set up and activated before the DNS query.
Otherwise, the query will fail.

Waiting for an IPv6 connection
******************************

.. code-block:: c

   at_fd = nrf_socket(NRF_AF_LTE, 0, NRF_NPROTO_AT);
   nrf_send(at_fd, "AT+CGEREP=1", strlen("AT+CGEREP=1"));
   # -> AT reponse "OK"

   # Set up PDN socket using the APN (PDN initialize and connect)

   nrf_getsockopt(pdn_fd, NRF_SOL_PDN, NRF_SO_PDN_CONTEXT_ID, &context_id, sizeof(uint8_t));
   while not (context_id in ":CGEV IPV6 <cid>");
   nrf_close(at_fd);

> When requesting an IPv6 connection (which requires PDN type IPV6 or IPV4V6), the IPv6 address might not be correct even if the request is successful.

Reviewer: This paragraph could precede the code snippet and the one below immediately follow it

The reason for this is that many networks use SLAAC, where the prefix of the IPv6 address is set after the PDN connection is created.
Therefore, you might need to wait with using an IPv6 socket until the prefix is properly set.
If the address cannot be obtained, a ``+CGEV: IPV6 FAIL <cid>`` error is returned.

The example code shows how to set up an AT socket that logs "CGEV" entries.
It signals when IPv6 is available for use by an IPv6 socket, either while connecting to a PDN or afterwards.

Note that the example code does not cover the default PDN (CID 0), because this PDN is always available.

> But the algorithm for making sure that IPv6 is available on the default PDN would be the same as if it was manually set up with an APN.

Reviewer: ```suggestion
However, the algorithm for checking that IPv6 is available on the default PDN is the same as if it was manually set up with an APN.
```

To do so, set the socket option NRF_SO_PDN_AF to request a specific type of connection.
You must do this after creating the socket and before activating it with ``nrf_connect``.

To retrieve a list of what was activated by the network, call ``getsockopt`` again after activation, using the same socket option.

.. admonition:: Limitation

   The current implementation is hardcoded to always request a dual IP stack.

Checking if a PDN is active
***************************

.. code-block:: c

   nrf_pdn_state_t state;
   nrf_getsockopt(pdn_fd, NRF_SOL_PDN, NRF_SO_PDN_STATE, &state, sizeof(nrf_pdn_state_t));

If a data socket is bound to an APN/PDN that is closed, the socket fails with an ``NRF_ENETDOWN`` error.

The example code shows how to check whether a PDN socket is still active.
However, if there is a long time between transmissions, the PDN socket might appear to still be active, even though the network has deactivated the PDN.
Only checking the state of the PDN by querying the socket option NRF_SO_PDN_STATE is therefore not reliable.
Instead, send a transmission on the data socket to trigger the link to become active again.

> The network will then report if the PDN has been deactivated.

Reviewer 2: I would rephrase this a bit, perhaps:
```
The example code shows how to check whether a PDN socket is still active.
However, if there is a long time between transmissions, the PDN socket state might be reported as active, even though the network has deactivated the PDN. In this case, sending data on the socket will fail with error NRF_ENETDOWN.
```

Author: Updated, but do we then need l. 98 (which says the exact same thing I think)?
And should we add something about that you should check the error code?

Reviewer 2: Yes, those two lines are redundant.

Author: OK, removed the first one.
